### PR TITLE
8314694: Separate checked_cast from globalDefinitions.hpp

### PR DESCRIPTION
**Pull request by kimbarrett**
Please review this change which moves checked_cast from globalDefinitions.hpp
to a separate file.  As part of this change we modify files that use
checked_cast to directly include that new file.  There are around 80 such
files, and that change constitutes the majority of the changed files and lines
in this PR.

This PR doesn't fix the definition of checked_cast (see JDK-8314258).  It just
moves the existing definition to a new file, in preparation for fixing it
later.  (I'm running tests on a fixed implementation.)

An alternative is to move checked_cast to a new file but have
globalDefinitions.hpp include that new file.  This would avoid touching the
include lists of currently using files.  It seems to me better to actually
separate it.

Fortunately, there was only one copyright update needed. Most of the uses were
added recently as part of addressing -Wconversion warnings, so those files
have already had copyright updates recently.

The other change was to move pointer_delta_as_int next to the related
pointer_delta, and change it to use a direct assert and static_cast, rather
than checked_cast.

With the exception of the simple change to pointer_delta_as_int the changes
in this PR are very simple and almost mechanical.  To find the files needing
an additional include and to demonstrate completing that task, I applied this
command to the hotspot directory:

```
egrep -r --files-with-matches --exclude-dir=.git " checked_cast<" . | \
  xargs egrep --files-without-match "utilities/checkedCast\.hpp"
```

So perhaps this change is trivial, despite the number of files.

Testing:
mach5 tier1

/labels hotspot

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8314694](https://bugs.openjdk.org/browse/JDK-8314694): Separate checked_cast from globalDefinitions.hpp (**Enhancement** - P4)

### Reviewers
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to [261d5c02](https://git.openjdk.org/jdk/pull/15377/files/261d5c02bd645f9723882d88e2b42c73ead64275)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [261d5c02](https://git.openjdk.org/jdk/pull/15377/files/261d5c02bd645f9723882d88e2b42c73ead64275)
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15377/head:pull/15377` \
`$ git checkout pull/15377`

Update a local copy of the PR: \
`$ git checkout pull/15377` \
`$ git pull https://git.openjdk.org/jdk.git pull/15377/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15377`

View PR using the GUI difftool: \
`$ git pr show -t 15377`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15377.diff">https://git.openjdk.org/jdk/pull/15377.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15377#issuecomment-1687393786)

### CODE REVIEW

**Review by tstuefe — APPROVED** [review]
Looks good.

Interesting how many of the touched cpp files did not include globalDefinitions.hpp.

File: src/hotspot/share/utilities/globalDefinitions.hpp [hunk 1]
```diff
@@ -475,6 +475,16 @@ inline size_t pointer_delta(const MetaWord* left, const MetaWord* right) {
...
 }
 
+// pointer_delta_as_int is called to do pointer subtraction for nearby pointers that
+// returns a non-negative int, usually used as a size of a code buffer range.
+// This scales to sizeof(T).
+template <typename T>
+inline int pointer_delta_as_int(const volatile T* left, const volatile T* right) {
+  size_t delta = pointer_delta(left, right, sizeof(T));
+  assert(delta <= size_t(INT_MAX), "pointer delta out of range: %zu", delta);
+  return static_cast<int>(delta);
```
Comment (coleenp) [inline src/hotspot/share/utilities/globalDefinitions.hpp:485]: I see what you did here...

**Review by coleenp — APPROVED** [review]
Looks good!

**Comment by kimbarrett** [comment]
Thanks for reviews @tstuefe and @coleenp .

> Interesting how many of the touched cpp files did not include globalDefinitions.hpp.

Oh, there are much worse things than that!  Take a look at utilities/elfFile.hpp :(

There was some exploration of the "Include What You Use" tool a long time ago, but it seemed like it didn't quite do what
we wanted.  https://include-what-you-use.org/

I'll run the command to check for using files that are missing the new include before integration.  That might show more
files needing the include added.  For example, https://github.com/openjdk/jdk/pull/15289 is ready to integrate.

**Review by sspitsyn — APPROVED** [review]
Changes in `src/hotspot/share/prims` folder look good.
Thanks,
Serguei

**Pull request closed without merging**